### PR TITLE
fix: improve ChannelList performance by improving threshold calls and rerenders of Header

### PR DESCRIPTION
**Pull request by khushal87**
## 🎯 Goal

This PR focuses on fixing the Header rerender issue in ChannelList if provided, and multiple `queryChannels` call when the threshold is reached and more data is loaded.

## 🛠 Implementation details

To fix the Header rerender the null check and `channels.length` check is removed as it wasn't really needed.
To fix the threshold API hits https://github.com/facebook/react-native/issues/14015#issuecomment-310675650 and  https://stackoverflow.com/a/60666252/10826415 was followed.

## 🎨 UI Changes

<details>
<summary>iOS</summary>

<table>
    <thead>
        <tr>
            <td>Before</td>
            <td>After</td>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                
            </td>
            <td>
                
            </td>
        </tr>
    </tbody>
</table>
</details>

<details>
<summary>Android</summary>

<table>
    <thead>
        <tr>
            <td>Before</td>
            <td>After</td>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                
            </td>
            <td>
                
            </td>
        </tr>
    </tbody>
</table>
</details>

## ☑️ Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] PR targets the `develop` branch
- [x] Documentation is updated
- [x] New code is tested in main example apps, including all possible scenarios
  - [x] SampleApp iOS and Android
  - [x] Expo iOS and Android

### CODE REVIEW

File: package/src/components/ChannelList/ChannelListMessenger.tsx [hunk 1]
```diff
@@ -151,14 +152,12 @@ const ChannelListMessengerWithContext = <
   }
 
   const onEndReached = () => {
-    if (loadNextPage) {
+    if (!onEndReachedCalledDuringMomentumRef.current && hasNextPage) {
       loadNextPage();
+      onEndReachedCalledDuringMomentumRef.current = true;
```
Comment (santhoshvai) [inline package/src/components/ChannelList/ChannelListMessenger.tsx:157]: A suggestion, can we rename `onEndReachedCalledDuringMomentumRef` to `onEndReachedCalledDuringCurrentScrollRef`... better understandable I guess

File: package/src/components/ChannelList/hooks/usePaginatedChannels.ts [hunk 2]
```diff
@@ -47,19 +47,21 @@ export const usePaginatedChannels = <
   setForceUpdate,
   sort = {},
 }: Parameters<StreamChatGenerics>) => {
-  const { client } = useChatContext<StreamChatGenerics>();
   const [channels, setChannels] = useState<Channel<StreamChatGenerics>[] | null>(null);
+  const [error, setError] = useState<Error | undefined>(undefined);
   const [staticChannelsActive, setStaticChannelsActive] = useState<boolean>(false);
+  const [activeQueryType, setActiveQueryType] = useState<QueryType | null>(null);
```
Comment (santhoshvai) [inline package/src/components/ChannelList/hooks/usePaginatedChannels.ts:53]: why the change of initial state to null from "queryLocalDB"

initially we must query the local db.. crucial when there is no network and offline.. has this been tested?

Comment (khushal87) [reply package/src/components/ChannelList/hooks/usePaginatedChannels.ts:53]: Just a un-noticed mistake, let me revert back

File: package/src/components/ChannelList/hooks/usePaginatedChannels.ts [hunk 3]
```diff
@@ -47,19 +47,21 @@ export const usePaginatedChannels = <
...
+  const [hasNextPage, setHasNextPage] = useState<boolean>(false);
```
Comment (santhoshvai) [inline package/src/components/ChannelList/hooks/usePaginatedChannels.ts:54]: also here, initally hasnextPage is true right as before? we dont know if its false.. s

Comment (khushal87) [reply package/src/components/ChannelList/hooks/usePaginatedChannels.ts:54]: It makes more sense to be false initially. If there's more data it will be true

File: package/src/components/ChannelList/hooks/usePaginatedChannels.ts [hunk 4]
```diff
@@ -88,7 +90,6 @@ export const usePaginatedChannels = <
     filtersRef.current = filters;
     sortRef.current = sort;
     isQueryingRef.current = true;
-    setError(undefined);
```
Comment (santhoshvai) [inline package/src/components/ChannelList/hooks/usePaginatedChannels.ts:91]: why is this removed? on every query removing the error seems fine to me..

Comment (santhoshvai) [reply package/src/components/ChannelList/hooks/usePaginatedChannels.ts:91]: if its causing extra rerender.. we have to merge the state somehow

Comment (khushal87) [reply package/src/components/ChannelList/hooks/usePaginatedChannels.ts:91]: There's already a same log on line 136

Comment (khushal87) [reply package/src/components/ChannelList/hooks/usePaginatedChannels.ts:91]: I don't understand why do we need it twice, if you can help here?

File: package/src/components/ChannelList/hooks/usePaginatedChannels.ts (hunk 3 above)
Comment (santhoshvai) [reply package/src/components/ChannelList/hooks/usePaginatedChannels.ts:54]: Usually these kind of things are set to true initially.. also explained by vishal in the call

File: package/src/components/ChannelList/hooks/usePaginatedChannels.ts (hunk 4 above)
Comment (santhoshvai) [reply package/src/components/ChannelList/hooks/usePaginatedChannels.ts:91]: Its better to remove the line 136 and keep this actually.. 

this one is essential.. removing any error state before making a request, 136 removes any error if request was successful.. both places is also fine, if state is same it wont rerender

File: package/src/components/ChannelList/hooks/usePaginatedChannels.ts (hunk 3 above)
Comment (khushal87) [reply package/src/components/ChannelList/hooks/usePaginatedChannels.ts:54]: But why can't it be false, this will never be true for the first fetch and can be true only if there are more data to load, right? Just asking to check if I am wrong here.

Eg: See `loadingExtraData` here - https://alexb72.medium.com/how-to-add-pagination-to-a-react-native-flatlist-ce2425d02f1a

File: package/src/components/ChannelList/hooks/usePaginatedChannels.ts (hunk 4 above)
Comment (khushal87) [reply package/src/components/ChannelList/hooks/usePaginatedChannels.ts:91]: Removing the 2nd one then

File: docusaurus/docs/reactnative/common-content/core-components/channel-list/props/load_more_threshold.mdx [hunk 5]
```diff
@@ -1,5 +1,5 @@
-Sets the [`onEndReachedThreshold`](https://reactnative.dev/docs/flatlist#onendreachedthreshold) of the underlying FlatList.
+Sets the [`onEndReachedThreshold`](https://reactnative.dev/docs/flatlist#onendreachedthreshold) of the underlying FlatList. We recommend using `0.1` as the default value for this prop as changing it might hit additional `channelQuery` calls and you might hit the limit.
```
Comment (vanGalilea) [inline docusaurus/docs/reactnative/common-content/core-components/channel-list/props/load_more_threshold.mdx:1]:
```suggestion
Sets the [`onEndReachedThreshold`](https://reactnative.dev/docs/flatlist#onendreachedthreshold) of the underlying FlatList. We recommend using `0.1` as the default value for this prop, as changing it might hit additional `channelQuery` calls, and you might reach the limit.
```

Comment (vanGalilea) [reply docusaurus/docs/reactnative/common-content/core-components/channel-list/props/load_more_threshold.mdx:1]: 2 times hit so I've changed it.

File: package/src/components/ChannelList/hooks/usePaginatedChannels.ts (hunk 3 above)
Comment (vanGalilea) [reply package/src/components/ChannelList/hooks/usePaginatedChannels.ts:54]: Aren't we receiving that info from our BE after the first fetch? I thought we had a "page_count" or such?

**Comment by vanGalilea** [comment]
This PR get's my approval after we settle the discussion about the initial state of `hasNextPage`

File: package/src/components/ChannelList/hooks/usePaginatedChannels.ts (hunk 3 above)
Comment (khushal87) [reply package/src/components/ChannelList/hooks/usePaginatedChannels.ts:54]: This is calculated using `channelQueryResponse.length >= newOptions.limit` everytime and also initially, so even if it's false, if there's more data it will be true. I don't understand what is wrong in this, correct me if I am wrong @vanGalilea

Comment (khushal87) [reply package/src/components/ChannelList/hooks/usePaginatedChannels.ts:54]: Technically it makes more sense to keep it false so that no logic will load more data initially no matter there's any data or not, if there's more data it will be true and the query will be called automatically

Comment (santhoshvai) [reply package/src/components/ChannelList/hooks/usePaginatedChannels.ts:54]: if its false initially.. why does it query? I thought we query only if hasNextPage is true

Comment (khushal87) [reply package/src/components/ChannelList/hooks/usePaginatedChannels.ts:54]: If you go through the code. `queryType` is `reload` for the first time, which means it should call the first 30 sets of data, then if there's more data or equal data we set the `hasNextPage` to true and next call is done to get more data. Initially if it's `false` it will be true after the first call if there's more data. What's the point in keeping it `true` since the beginning?

Comment (khushal87) [reply package/src/components/ChannelList/hooks/usePaginatedChannels.ts:54]: Check the `useEffect` for !enableOfflineSupport case and you will get what is happening.

Comment (vanGalilea) [reply package/src/components/ChannelList/hooks/usePaginatedChannels.ts:54]: @santhoshvai we set the `hasNextPage` only if after the first `queryChannels` call  is made and when there is an extra page/data (we do it by calculating `channelQueryResponse.length >= newOptions.limit`).

Wit this PR we call `queryChannels` only when needed (hasNextPage is true) and use of resources has been limited

**Review by vanGalilea — APPROVED** [review]

**Comment by vanGalilea** [comment]
After Beta release let's share with the relevant integrators!

**Comment by vanGalilea** [comment]
https://www.npmjs.com/package/stream-chat-react-native/v/5.8.1-beta.1
